hal/halx86/smp/i386/spinup.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     i386 AP spinup setup
+ * COPYRIGHT:   Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ *              Copyright 2021 Victor Perevertkin <victor.perevertkin@reactos.org>
+ */
+
+/* INCLUDES ******************************************************************/
+
+#include <hal.h>
+#include <smp.h>
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS *******************************************************************/
+
+extern PPROCESSOR_IDENTITY HalpProcessorIdentity;
+extern PHYSICAL_ADDRESS HalpLowStubPhysicalAddress;
+extern PVOID HalpLowStub;
+
+// The data necessary for a boot (stored inside HalpLowStub)
+extern PVOID HalpAPEntry16;
+extern PVOID HalpAPEntryData;
+extern PVOID HalpAPEntry32;
+extern PVOID HalpAPEntry16End;
+extern HALP_APIC_INFO_TABLE HalpApicInfoTable;
+
+ULONG HalpStartedProcessorCount = 1;
+
+typedef struct _AP_ENTRY_DATA
+{
+    ULONG Jump32Offset;
+    ULONG Jump32Segment;
+    PVOID SelfPtr;
+    PFN_NUMBER PageTableRoot;
+    PKPROCESSOR_STATE ProcessorState;
+    KDESCRIPTOR Gdtr;
+} AP_ENTRY_DATA, *PAP_ENTRY_DATA;
+
+typedef struct _AP_SETUP_STACK
+{
+    PVOID ReturnAddr;
+    PVOID KxLoaderBlock;
+} AP_SETUP_STACK, *PAP_SETUP_STACK;
+
+/* FUNCTIONS *****************************************************************/
+
+static
+VOID
+HalpMapAddressFlat(
+    _Inout_ PMMPDE PageDirectory,
+    _In_ PVOID VirtAddress,
+    _In_ PVOID TargetVirtAddress)
+{
+    if (TargetVirtAddress == NULL)
+        TargetVirtAddress = VirtAddress;
+
+    PMMPDE currentPde;
+
+    currentPde = &PageDirectory[MiAddressToPdeOffset(TargetVirtAddress)];
+
+    // Allocate a Page Table if there is no one for this address
+    if (currentPde->u.Long == 0)
+    {
+        PMMPTE pageTable = ExAllocatePoolZero(NonPagedPoolMustSucceed , PAGE_SIZE, TAG_HAL);
+
+        currentPde->u.Hard.PageFrameNumber = MmGetPhysicalAddress(pageTable).QuadPart >> PAGE_SHIFT;
+        currentPde->u.Hard.Valid = TRUE;
+        currentPde->u.Hard.Write = TRUE;
+    }
+    
+    // Map the Page Table so we can add our VirtAddress there (hack around I/O memory mapper for that)
```

Comment:
In which way do we "hack around" that? What do we do as a workaround? And what should actually be done instead?

Reply:
Getting the technical explanation written here with words was my goal. Resolved. "Doing it properly" is by definition == "Doing it as Windows does it".